internal/docker/docker.go

```
@@ -0,0 +1,177 @@
+	"github.com/konstructio/colony/internal/logger"
+)
+
+type Client struct {
+	cli *client.Client
+	log *logger.Logger
+}
+
+func New(logger *logger.Logger) (*Client, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, fmt.Errorf("error creating docker client: %w", err)
+	}
+
+	return &Client{
+		cli: cli,
+		log: logger,
+	}, nil
+}
+
+// func getColonyK3sContainerIDAndName(ctx context.Context, c *Client) {
+func getColonyK3sContainerIDAndName(ctx context.Context, c *Client) (string, string, error) {
+
+	containers, err := c.cli.ContainerList(ctx, containerTypes.ListOptions{All: true})
+	if err != nil {
+		return "", "", fmt.Errorf("error listing containers on host: %w", err)
+	}
+
+	for _, container := range containers {
+		if container.Names[0] == "/"+constants.ColonyK3sContainerName {
+			return container.ID, container.Names[0], nil
+		}
+	}
+	return "", "", fmt.Errorf("container %s not found", constants.ColonyK3sContainerName)
+}
+
+func (c *Client) RemoveColonyK3sContainer(ctx context.Context) error {
+
+	defer c.cli.Close()
+
+	colonyK3sContainerID, colonyK3sContainerName, err := getColonyK3sContainerIDAndName(ctx, c)
+	if err != nil {
+		return fmt.Errorf("error getting %s container: %v", constants.ColonyK3sContainerName, err)
+	}
+	c.log.Info(fmt.Sprintf("found container name %s with ID %s  ", strings.TrimPrefix(colonyK3sContainerName, "/"), colonyK3sContainerID[:constants.DefaultDockerIDLength]))
+
+	err = c.cli.ContainerRemove(ctx, colonyK3sContainerID, containerTypes.RemoveOptions{Force: true})
+	if err != nil {
+		return fmt.Errorf("error removing container: %v", err)
+	}
+	return nil
+}
+
+func (c *Client) CreateColonyK3sContainer(ctx context.Context) error {
+	log := logger.New(logger.Debug)
+
+	defer c.cli.Close()
+
+	// check for an existing colony-k3s container
+	colonyK3sContainerID, _, _ := getColonyK3sContainerIDAndName(ctx, c)
```

Wouldn't this eat any error like the docker daemon being in a bad state?

I feel we would want to know those things.